src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -323,7 +340,9 @@ Object invokeMoves(Object[] args, Binding.Move[] argBindings, Binding.Move[] ret
     Object invokeInterpBindings(Object[] args, MethodHandle leaf,
                                 Map<VMStorage, Integer> argIndexMap,
                                 Map<VMStorage, Integer> retIndexMap) throws Throwable {
-        NativeScope scope = bufferCopySize != 0 ? NativeScope.boundedScope(bufferCopySize) : null;
+        boolean hasBufferCopies = bufferCopySize != 0;
+        NativeScope scope = hasBufferCopies ? NativeScope.boundedScope(bufferCopySize) : null;
```

Comment:
If Allocator had a close() method, this could be simplified a little?

Reply:
This is not really a problem with Allocator, it's a problem with having either a NativeScope, or having none (null), which will result in an NPE when trying to reference `scope::allocateNative`. That's why this code is conditional.

Reply:
I'm experimenting with this a bit. Seems like there is some room for improvement.